talk/expert/cpp20spaceship.tex

```diff
@@ -0,0 +1,242 @@
+    \end{itemize}
+  \end{block}
+\end{frame}
+
+\begin{frame}[fragile]
+  \frametitlecpp[20]{The three-way comparison operator}
+  \begin{block}{Default {\it <=>} implementation}
+    \begin{itemize}
+      \item One can ask the compiler to provide some default implementation for {\it <=>}. Logically enough, it will compare the first member variable of the two objects, and goes on to the next member variable as long as the current ones are equivalent.
+      \item In the previous example, that would not be relevant, because it will compare the numerators first, and conclude that {\it 3/6} is greater than {\it 2/3}:
+    \end{itemize}
+  \end{block}
+  \begin{exampleblock}{Example}
+    \begin{cppcode*}{}
+#include <iostream>
+
+struct Rational
+ {
+  unsigned n, d ;
+  std::strong_ordering operator<=>( Rational const & other ) const = default ;
+ } ;
+
+std::ostream & operator<<( std::ostream & os, std::weak_ordering cmp )
+ { return (os<<'<'<<(cmp<0)<<'|'<<(cmp==0)<<'|'<<(cmp>0)<<'>') ; }
+
+std::ostream & operator<<( std::ostream & os, Rational const & r )
+ { return (os<<r.n<<'/'<<r.d) ; }
+
+int main()
+ {
+  Rational r1 { 3, 6 }, r2 { 2, 3 } ;
+  std::cout<<r1<<" <=> "<<r2<<": "<<(r1<=>r2)<<std::endl ;
+ }
+    \end{cppcode*}
+  \end{exampleblock}
+\end{frame}
+
+\begin{frame}[fragile]
+  \frametitlecpp[20]{The three-way comparison operator}
+  \begin{block}{Default {\it <=>} implementation}
+    \begin{itemize}
+    \item On the contrary, for some tuple-like class, it may make sense.
+    \end{itemize}
+  \end{block}
+  \begin{exampleblock}{Example}
+    \begin{cppcode*}{}
+#include <iostream>
+#include <string>
+#include <set>
+
+struct Grade
+ {
+  double number ;
+  char letter ;
+  std::string name ;
+
+  auto operator<=>( Grade const & other ) const = default ;
+ } ;
+
+std::ostream & operator<<( std::ostream & os, Grade const & g )
+ { return (os<<g.letter<<", "<<g.number<<", "<<g.name) ; }
+
+int main()
+ {
+  std::set<Grade> grades
+   {
+     { 19, 'A', "Djamila" },
+     { 12, 'C', "Charles" },
+     { 16.5, 'A', "Marc" },
+   } ;
+
+  for ( auto const & grade : grades )
+   { std::cout<<grade<<std::endl ; }
+ }
+    \end{cppcode*}
+  \end{exampleblock}
+\end{frame}
+
+\begin{frame}[fragile]
+  \frametitlecpp[20]{The three-way comparison operator}
+  \begin{block}{Default {\it <=>} implementation}
+    \begin{itemize}
+      \item We see above that the definition of {\it <=>} has been completed by the compiler with a default implementation of {\it <} (among others), which has been used by {\it std::set} so to sorts its elements.
+      \item Worth to note : if you use the \textbf{default} implementation of {\it <=>}, you will also get additional {\it ==} and {\it !=}. Not sure this special behavior is such a good idea...
```

Comment:
I would not judge this design decision and just stick to the facts.
```suggestion
      \item If {\it <=>} is defaulted and no {\it ==} is defined, then the compiler also provides a defaulted {\it ==}.
```

Reply:
My "worth" was not meant to say it is a good design choice ;)